core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -187,6 +187,32 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
   void execute() const {
     ReducerTypeFwd functor = ReducerConditional::select(m_functor, m_reducer);
 
+    if (m_policy.begin() == m_policy.end()) {
+      const Kokkos::Experimental::SYCL& space = m_policy.space();
+      Kokkos::Experimental::Impl::SYCLInternal& instance =
+          *space.impl_internal_space_instance();
+      cl::sycl::queue& q = *instance.m_queue;
+
+      sycl::usm::alloc result_ptr_type =
+          sycl::get_pointer_type(m_result_ptr, q.get_context());
+
+      switch (result_ptr_type) {
+        case sycl::usm::alloc::host:
+        case sycl::usm::alloc::shared:
+          ValueInit::init(functor, m_result_ptr);
+          break;
+        case sycl::usm::alloc::device:
+        case sycl::usm::alloc::unknown:
+          value_type host_result;
+          ValueInit::init(functor, &host_result);
+          q.memcpy(m_result_ptr, &host_result, sizeof(host_result));
+          break;
+        default: break;
+      }
+      q.wait();
```

Comment:
We can move the `q.wait()` to right after the `q.memcpy()`, or even `q.memcpy(...).wait()`, as that is the only place it is used.

Reply:
Minor:  add a comment about the "unknown" memory, because it really is non-USM-allocated memory (for lack of a better term, since SYCL can have host and CPU devices which would map to USM host memory).